### PR TITLE
Add location alias

### PR DESCRIPTION
**Pull request by RenCloud**
So this is the PR for #242. Hopefully i have not missed something. 
Adds the command `friends add alias`. Quite similar to `friends add nickname`.
Of course it also adds the counter part `friends remove alias`.

I have also added some test cases based on the nickname test cases.

`alias` is unfortunately a keyword in ruby, so i named the variables `nickname` . While function and description using `alias` to match the command. 

---

Hi there! Thanks so much for submitting a pull request!

Let's just make sure together that all of these boxes are checked before we
merge this change:

- [x] The code in these changes works correctly.
- [x] Code has tests as appropriate.
- [x] Code has been reviewed by @JacobEvelyn.
- [x] All tests pass on Travis CI.
- [x] Code coverage remains high.
- [x] Rubocop reports no issues on Travis CI.
- [x] The `README.md` file is updated as appropriate.

Don't worry—this list will get checked off in no time!

### CODE REVIEW

**Comment by RenCloud** [comment]
Oh i just saw that i print already everything with `friends list locations` while `friends list friends` only list nicknames and tag with the parameter `--verbose` . I will adopt the behavior to `friends list locations`.

File: README.md
Comment (JacobEvelyn) [inline README.md:296]: Could you change this to `or their aliases`?

File: README.md
Comment (JacobEvelyn) [inline README.md:894]: Could you change `an` to `a` here?

File: README.md
Comment (JacobEvelyn) [inline README.md:301]: I love this example. Could you change `NYC` to `nyc` so we can also show that it matches case-insensitively as we do with `atlantis`?

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [inline lib/friends/introvert.rb:218]: Could you change `location` to `locations` here?

File: lib/friends/commands/remove.rb
Comment (JacobEvelyn) [inline lib/friends/commands/remove.rb:18]: I noticed that this (and the same with adding/removing nicknames) doesn't use `.to_s` but the add does. I assume that's to handle `nil` values, which makes me wonder how this behaves with no arguments passed (or only one). Does it break? Are the `.to_s` calls not needed? Or am I missing something?

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [inline lib/friends/introvert.rb:224]: Thanks for adding this check! A few small comments:

1. Since we'll want to know which location matches, can we do the lookup once instead of do `any?` and then the `select`?
2. Can we also check that no location alias matches another location's _name_? (And similarly add that check in `add_location`.)
3. `find` is a more efficient alternative to `select { ... }.first`
4. Should we do this matching case-insensitively? (I realize that `add_location` and `add_friend` don't currently use case-insensitive matching but I should probably change that.) I'm happy to move that to a future PR if you'd prefer.

Concretely, I'm imagining something like:

```ruby
collision = @locations.find do |loc|
  loc.name.casecmp(nickname) == 0 || loc.nicknames.any? { |n| n.casecmp(nickname) == 0 }
end

if collision
  raise FriendsError, ...
end
```

And actually if we're doing this check in both `add_location` and `add_location_alias` we should probably extract it to a helper method.

Thoughts?

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [inline lib/friends/introvert.rb:220]: Would you mind changing these method names to `add_alias` and `remove_alias` to more closely match the CLI commands?

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [inline lib/friends/introvert.rb:277]: Same typo: `locations`

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [inline lib/friends/introvert.rb:432]: Eek, good catch!

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [inline lib/friends/introvert.rb:440]: If we're only ever going to match one location and not be smart about it, can we keep this as a straight map?

File: lib/friends/location.rb
Comment (JacobEvelyn) [inline lib/friends/location.rb:14]: Small typo: `REFIX` -> `PREFIX`

I'd actually prefer to use `alias` everywhere (e.g. `ALIAS_PREFIX`, `alias_str`, `@aliases`, etc.) except where we have to use `nickname` because of the keyword issue. Thoughts?

File: lib/friends/location.rb
Comment (JacobEvelyn) [inline lib/friends/location.rb:19]: I'm not sure I understand all of this—what is the `@` stuff for?

File: lib/friends/location.rb
Comment (JacobEvelyn) [inline lib/friends/location.rb:38]: Ooh this is clever! Would you mind changing `self` to `to_s` to make it a little clearer what's going on?

File: lib/friends/location.rb
Comment (JacobEvelyn) [inline lib/friends/location.rb:53]: Small typo: `Add an alias`

File: lib/friends/location.rb
Comment (JacobEvelyn) [inline lib/friends/location.rb:79]: This comment seems incorrect?

File: lib/friends/location.rb
Comment (JacobEvelyn) [inline lib/friends/location.rb:76]: I find this comment a bit confusing because the following line isn't what's doing the "returning." Is there a way to reword it?

File: lib/friends/location.rb
Comment (JacobEvelyn) [inline lib/friends/location.rb:77]: I think this allow-any-amount-of-whitespace thing is fine but not actually necessary. (And honestly it was probably unnecessary for friend regexes too.) I'm wondering if something like:

```ruby
[name, @nicknames].map { |str| Friends::RegexBuilder.regex(str) }
```

is really what we want and simplifies a lot?

File: test/commands/add/alias_spec.rb
Comment (JacobEvelyn) [inline test/commands/add/alias_spec.rb:44]: Would you mind adding another test for when `nickname` is just `nil`? That will I think capture the `friends add nickname 'New York City'` example that I want to make sure we can handle.

File: test/commands/add/alias_spec.rb
Comment (JacobEvelyn) [inline test/commands/add/alias_spec.rb:60]: Could you change `friend` here to `location`?

Comment (JacobEvelyn) [inline test/commands/add/alias_spec.rb:60]: Actually, I'm not sure I understand how this test is different from the one above. What am I missing?

File: test/commands/remove/alias_spec.rb
Comment (JacobEvelyn) [inline test/commands/remove/alias_spec.rb:6]: Could you also add tests for when `location_name` and `nickname` are `nil`, and when just `nickname` is?

**Review by JacobEvelyn — CHANGES_REQUESTED** [review]
Thanks so much for all the hard work, @RenCloud! This generally looks good.

I added some comments; please let me know if you have questions about or disagree with anything—I'm always open to discussing!

File: lib/friends/commands/remove.rb
Comment (RenCloud) [reply lib/friends/commands/remove.rb:18]: Oh, that happens when you work to much based on the more similar command 😅. You're correct. The add methods used it and provide better error handling than all of the remove methods. `remove tag` also seems to avoid the additional check.   

It didn't seems to directly break something, but it handles the missing input not as good as the `add alias` method. This could be the reason that it was not noticed before.

For the case that both arguments are missing, somehow the error from `regex.match(nil)` is catch somewhere. At least it doesn't kill the method `thing_with_name_in`. For this case i'm not sure why it works quite normal. 

If the location is given, but not the alias than the `include?` check will fail.  

----



I will add the two checks, like in the `add alias` function
```ruby
raise FriendsError, "Expected \"[Location Name]\" \"[Alias]\"" if name.empty?
raise FriendsError, "Alias cannot be blank" if nickname.empty?
```
in addition to the `.to_s` call to create the same behavior as with the `add alias` method.

Similar that should be done for nickname. Also for the additional tests for that behavior (no argument given, only friends name given).

File: lib/friends/introvert.rb
Comment (RenCloud) [reply lib/friends/introvert.rb:220]: Right, i forgot to change their names after the discussion about how the command should be named.

File: lib/friends/location.rb
Comment (RenCloud) [reply lib/friends/location.rb:19]: That's right. The `@` stuff is not needed also `\[` is not needed. Because i first failed to create the right regex, i build it with the help of the deserialization_regex of friends.

So it is a bit shorter `[^\(]*[^\(\s]`.
Which then should be all character until ` (`. 

I tried a bit, but currently i didn't find a shorter solution. But i'm not so deep in regex may there is a way.

File: lib/friends/location.rb
Comment (RenCloud) [reply lib/friends/location.rb:77]: Yeah the whitespace thing is not really needed. 

It should do the same thing and then the comment match more the line. I will use it,

File: test/commands/add/alias_spec.rb
Comment (RenCloud) [reply test/commands/add/alias_spec.rb:60]: You are right. This test didn't test something new. I will remove it.

File: lib/friends/introvert.rb
Comment (RenCloud) [reply lib/friends/introvert.rb:224]: 1. Yes, of course. The double checkup is not needed. 
2. Right, this could also happen and is a good point. 
3. I see. Also it is more descriptive for that case. Somehow only `find_index` was listed in the documentation i had open. 
4. The methods that match a name in an activity use a case-insensitiv regex. Because of that i think it would only make sense to do the check match while adding a new one case-insensitiv , too. I think another PR would be better. 


Yes after using that in both methods it should be exist a helper method.

That look great also seems to work as expected.

File: lib/friends/introvert.rb
Comment (RenCloud) [reply lib/friends/introvert.rb:440]: I'm not quite sure if i understand you here correct. We don't need something complex here, this is where I can follow you and agree.

So do you want something like before my changes, but that would also be a hash map.

Or only create a big map containing all names and alias. So something like
```ruby
def regex_location_map
  @locations.map { |location| [location.name, location.regex_for_name] }
end
```
Resulting in an `map[location_name, regex (with name and aliases)]`.

File: lib/friends/location.rb
Comment (RenCloud) [reply lib/friends/location.rb:38]: Yes, it would be a bit clearer, but robocop doesn't like that. Besides, `#{self}` is also used in the file `friend.rb` for the same purpose.

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [reply lib/friends/introvert.rb:440]: Sorry for being unclear! I was picturing something like before your changes (and I realize that my `# @return` comment was also wrong 😬 ), basically giving us something like `{ location_name => regex }`. You're right that the way we use this it doesn't actually need to be a hash and could just be an array (which `map` returns) since we're just iterating through them all, but I slightly prefer the feeling of the hash to show how the data's related.

Does that make sense?

File: lib/friends/location.rb
Comment (JacobEvelyn) [reply lib/friends/location.rb:38]: Ah I didn't realize RuboCop complained and that we used this elsewhere. This is fine to stay the way you have it then!

File: lib/friends/location.rb
Comment (JacobEvelyn) [reply lib/friends/location.rb:19]: So the way I think about this is we want the line in `friends.md` to match one of the following:

```
- New York City
- New York City (a.k.a. NYC)
- New York City (a.k.a. NYC a.k.a. the Big Apple a.k.a. Gotham)
```

to me that means it has the following:

- `- ` (the serialization prefix)
- zero or more characters that's not a `(`
- a last character in the name that's not a `(` or ` `
- (optionally) ` (` followed by `a.k.a. ` (the alias prefix) followed by one or more characters and then `)`

which I think would translate to:

```ruby
/(#{SERIALIZATION_PREFIX})?(?<name>[^\(]*[^\(\s])(\s+\(#{ALIAS_PREFIX}(?<alias_str>.+)\))?/
```

so in `initialize`, `name` would be `New York City` and `alias_str` would be `nil`, `NYC`, or `NYC a.k.a. the Big Apple a.k.a. Gotham`

File: lib/friends/introvert.rb
Comment (RenCloud) [reply lib/friends/introvert.rb:440]: No problem maybe i only understand it not correct.  Hope that i did now. I changed it back to a similar hash like before with `{location_name => Regexlist}`. It make sense yes. 

Also i changed the comment. Should now match more to the method.

Happens to the best 😃 .

File: lib/friends/location.rb
Comment (RenCloud) [reply lib/friends/location.rb]: Yes, i find that also better. `nickname` is not used so much so it should make no problem.

File: lib/friends/location.rb
Comment (RenCloud) [reply lib/friends/location.rb]: Yeah correct that works. My answer was a bit short. It referenced only the part i wanted to change, but not where it is.

**Comment by RenCloud** [comment]
Additional to the discussion in the comments, i fixed the typos and added the additional test cases. 

Unfortunately `String.casecmp?` does not exist in Ruby 2.3 so i changed it to `String.casecmp().zero?`.

File: lib/friends/introvert.rb
Comment (JacobEvelyn) [inline lib/friends/introvert.rb:480]: I'm so sorry—you changed this to exactly what I suggested but I made a typo—can we make this a hash of `{ regex => location }` instead to more closely match the `regex_friend_map`? That way, when we sort the keys by length, we know that `/Paris, France/` will be checked before `/Paris/` (which means we'd store activities as `Went shopping in _Paris, France_` instead of `Went shopping in _Paris_, France`).

Concretely I'm thinking something like:

```ruby
@locations.each_with_object({}) do |location, hash|
  location.regexes_for_name.each { |regex| hash[regex] = location }
end
```

Does that make sense?

**Review by JacobEvelyn — CHANGES_REQUESTED** [review]
Hi @RenCloud! Sorry for the delay—this looks perfect except for one tiny question and one small change that's completely my fault. (Sorry!)

Once that's changed this will be good to merge! 🎉 Let me know if you have any questions!

File: lib/friends/introvert.rb
Comment (RenCloud) [reply lib/friends/introvert.rb:231]: Yes, print also the alias could help. I will change that.

File: lib/friends/introvert.rb
Comment (RenCloud) [reply lib/friends/introvert.rb:480]: This makes sense to me. I change that, too

**Review by JacobEvelyn — APPROVED** [review]

**Comment by JacobEvelyn** [comment]
Thanks so much for all the hard work here, @RenCloud! I've just merged this and released it in v0.54. You can get it with `friends update`.

Congratulations on becoming an official [contributor](https://github.com/JacobEvelyn/friends/graphs/contributors), and I'd love to have you contribute again in the future!

**Pull request merged**
